File: AccuPay/Domain/Paystubs/PayRate.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema
Imports AccuPay.Data

Namespace Global.AccuPay.Entity

    <Table("payrate")>
    Public Class PayRate
        Implements IPayrate

        Public Const RegularDay As String = "Regular Day"

        Public Const SpecialNonWorkingHoliday As String = "Special Non-Working Holiday"

        Public Const RegularHoliday As String = "Regular Holiday"

        Public Const DoubleHoliday As String = "Double Holiday"

        Public Const RegularDayAndSpecialHoliday As String = "Regular + Special Holiday"

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer?

        Public Property OrganizationID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        Public Property DayBefore As Date?

        <Column("Date")>
        Public Property [Date] As Date

        Public Property PayType As String

        Public Property Description As String

        <Column("PayRate")>
        Public Property CommonRate As Decimal

        Public Property OvertimeRate As Decimal

        Public Property NightDifferentialRate As Decimal

        Public Property NightDifferentialOTRate As Decimal

        Public Property RestDayRate As Decimal

        Public Property RestDayOvertimeRate As Decimal

        Public Property RestDayNDRate As Decimal

        Public Property RestDayNDOTRate As Decimal

        Public ReadOnly Property IsRegularDay As Boolean Implements IPayrate.IsRegularDay
            Get
                Return PayType = RegularDay
            End Get
        End Property

        Public ReadOnly Property IsHoliday As Boolean Implements IPayrate.IsHoliday
            Get
                Return IsSpecialNonWorkingHoliday Or IsRegularHoliday
            End Get
        End Property

        Public ReadOnly Property IsSpecialNonWorkingHoliday As Boolean Implements IPayrate.IsSpecialNonWorkingHoliday
            Get
                Return PayType = SpecialNonWorkingHoliday
            End Get
        End Property

        Public ReadOnly Property IsRegularHoliday As Boolean Implements IPayrate.IsRegularHoliday
            Get
                Return PayType = RegularHoliday OrElse PayType = DoubleHoliday
            End Get
        End Property

        Public ReadOnly Property IPayrate_Date As Date Implements IPayrate.Date
            Get
                Return [Date]
            End Get
        End Property

        Public ReadOnly Property IPayrate_RegularRate As Decimal Implements IPayrate.RegularRate
            Get
                Return CommonRate
            End Get
        End Property

        Private ReadOnly Property IPayrate_OvertimeRate As Decimal Implements IPayrate.OvertimeRate
            Get
                Return OvertimeRate
            End Get
        End Property

        Public ReadOnly Property IPayrate_NightDiffRate As Decimal Implements IPayrate.NightDiffRate
            Get
                Return NightDifferentialRate
            End Get
        End Property

        Public ReadOnly Property IPayrate_NightDiffOTRate As Decimal Implements IPayrate.NightDiffOTRate
            Get
                Return NightDifferentialOTRate
            End Get
        End Property

        Private ReadOnly Property IPayrate_RestDayRate As Decimal Implements IPayrate.RestDayRate
            Get
                Return RestDayRate
            End Get
        End Property

        Public ReadOnly Property IPayrate_RestDayOTRate As Decimal Implements IPayrate.RestDayOTRate
            Get
                Return RestDayOvertimeRate
            End Get
        End Property

        Private ReadOnly Property IPayrate_RestDayNDRate As Decimal Implements IPayrate.RestDayNDRate
            Get
                Return RestDayNDRate
            End Get
        End Property

        Private ReadOnly Property IPayrate_RestDayNDOTRate As Decimal Implements IPayrate.RestDayNDOTRate
            Get
                Return RestDayNDOTRate
            End Get
        End Property

        Public ReadOnly Property IsDoubleHoliday As Boolean Implements IPayrate.IsDoubleHoliday
            Get
                Return PayType = DoubleHoliday
            End Get
        End Property

    End Class

End Namespace

File: Accupay.Data/Entities/PayRate.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("payrate")]
    public class PayRate : IPayrate
    {
        public const string RegularDay = "Regular Day";

        public const string SpecialNonWorkingHoliday = "Special Non-Working Holiday";

        public const string RegularHoliday = "Regular Holiday";

        public const string DoubleHoliday = "Double Holiday";

        public const string RegularDayAndSpecialHoliday = "Regular + Special Holiday";

        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int? RowID { get; set; }

        public int? OrganizationID { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public DateTime Created { get; set; }

        public int? CreatedBy { get; set; }

        [DatabaseGenerated(DatabaseGeneratedOption.Computed)]
        public DateTime? LastUpd { get; set; }

        public int? LastUpdBy { get; set; }

        public DateTime? DayBefore { get; set; }

        public DateTime Date { get; set; }

        public string PayType { get; set; }

        public string Description { get; set; }

        [Column("PayRate")]
        public decimal CommonRate { get; set; }

        public decimal OvertimeRate { get; set; }

        public decimal NightDifferentialRate { get; set; }

        public decimal NightDifferentialOTRate { get; set; }

        public decimal RestDayRate { get; set; }

        public decimal RestDayOvertimeRate { get; set; }

        public decimal RestDayNDRate { get; set; }

        public decimal RestDayNDOTRate { get; set; }

        public bool IsRegularDay => PayType == RegularDay;

        public bool IsHoliday => IsSpecialNonWorkingHoliday | IsRegularHoliday;

        public bool IsSpecialNonWorkingHoliday => PayType == SpecialNonWorkingHoliday;

        public bool IsRegularHoliday => PayType == RegularHoliday || PayType == DoubleHoliday;

        public decimal RegularRate => CommonRate;

        public decimal NightDiffRate => NightDifferentialRate;

        public decimal NightDiffOTRate => NightDifferentialOTRate;

        public decimal RestDayOTRate => RestDayOvertimeRate;

        public bool IsDoubleHoliday => PayType == DoubleHoliday;
    }
}